Inserter: Preview saved block on hover

youknowriad opened the pull request:
closes #4106

This PR tries to show a preview of the block to be inserted at the bottom of the inserter menu.
Ideas to improve the design are welcome.

I believe we could introduce a `sampleAttributes` property for blocks to make this preview more compelling using the sample attributes.

Maybe we could consider dropping the margins between the blocks in the inserter to avoid the flickering effect when moving from a block to another.

<img width="362" alt="screen shot 2018-01-03 at 13 58 50" src="https://user-images.githubusercontent.com/272444/34521483-4215ac66-f08e-11e7-8662-4f2cef19defe.png">

youknowriad committed:
Inserter: Preview block on hover

mtias commented:
> I believe we could introduce a sampleAttributes property for blocks to make this preview more compelling using the sample attributes.

My thought was an `example` property, modelled in a way that could be supplied to testing as well.

youknowriad committed:
Block API: Adding the "example" block property

aduth commented on editor/components/inserter/group.js:
```diff
@@ -56,6 +64,8 @@ export default class InserterGroup extends Component {
 				ref={ bindReferenceNode( block.name ) }
 				tabIndex={ current === block.name || disabled ? null : '-1' }
 				disabled={ disabled }
+				onMouseEnter={ this.hoverBlock( ( block ) ) }
```
Redundant inner parentheses?

aduth commented on editor/components/inserter/group.js:
```diff
@@ -39,6 +39,14 @@ export default class InserterGroup extends Component {
 		}
 	}
 
+	hoverBlock( block ) {
```
The function name doesn't seem entirely accurate. This is not the hover handler itself; it creates a hover handler. Maybe: `createToggleBlockHover` ?

Also: Maybe it's worthwhile to check `this.props.onHover` before returning anything here? That way, we could e.g. `return null;` and avoid creating new function references (i.e. forcing React to reconcile changing function handlers).

aduth commented on editor/components/inserter/style.scss:
```diff
@@ -1,3 +1,7 @@
+$content-height: 250px;
```
My understanding is these assigned variables are global to all stylesheets, and this seems like it might be a common name. Maybe worth prefixing to act as namespace if we want the variables?

youknowriad replied:
Good point

aduth commented on blocks/library/code/index.js:
```diff
@@ -32,6 +32,14 @@ registerBlockType( 'core/code', {
+	example: {
+		content: `
+<?php
+	echo "Beep Beeb Boop";
```
🤖

aduth replied:
Personally I'd find this reasonable to inline in the sense that even stylistically this is something I'd have on one line:

```js
example: {
	content: '<?php echo "Beep Beep Boop"; ?>',
},
```

aduth replied:
Typo: "Beeb" -> "Beep"

youknowriad committed:
Block Preview: Beep Beep Boop and styling fixes

youknowriad commented:
I added the `example` API and tried it for the "code" block only. 

 - Can I have some thoughts for this PR? Do you think I should remove the "margin" from the inserter items to avoid the flickering?

- What else can I do to polish this?

aduth commented on blocks/library/code/index.js:
```diff
@@ -32,6 +32,10 @@ registerBlockType( 'core/code', {
 	},
 
+	example: {
+		content: '<?php echo "Beep Beep Boop";?>',
```
Very minor: I'd expect a space before the closing `?>`

https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#opening-and-closing-php-tags

aduth replied:
Minor: PHP Guidelines recommend single quotes:

>If you’re not evaluating anything in the string, use single quotes.

https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#single-and-double-quotes

aduth commented on editor/components/block-preview/index.js:
```diff
@@ -0,0 +1,39 @@
+/**
+ * External dependencie
```
Typo: "dependencie" -> "dependencies"

aduth reviewed (commented):
The placement below the inserter doesn't work particularly well when the inserter opens downward (i.e. for new posts), where I'm seeing the preview clipping the bottom of the viewport and unable to be scrolled.

![Clipping](https://user-images.githubusercontent.com/1779930/34624073-dee71fd6-f221-11e7-8811-34536e907e3b.png)

mtias commented:
@jasmussen @youknowriad I think this might need a few more iterations, what do you think of enabling only for "saved blocks" initially? That provides immediate value and requires less consideration of dummy data since saved blocks, by their nature, don't need it.

jasmussen commented:
I think it's fine to start small and test. 👍 👍

youknowriad commented:
Should I remove the "example" API for now?

mtias commented:
> Should I remove the "example" API for now?

I don't mind retaining it but not documented.

youknowriad commented:
I updated this, it's now limited to the reusable blocks and I also updated slightly the design of the inserter "blocks" by removing the margin in favor of a padding to avoid the flickering when we move the mouse between saved blocks.

youknowriad commented:
Can I have a final review here?

jasmussen commented:
I think visually this looks pretty good. But I'm getting a few errors on hover:

<img width="1735" alt="screen shot 2018-01-15 at 10 30 33" src="https://user-images.githubusercontent.com/1204802/34935763-38702574-f9df-11e7-9da9-dc0fd713e3b3.png">

<img width="1735" alt="screen shot 2018-01-15 at 10 30 30" src="https://user-images.githubusercontent.com/1204802/34935768-39e173e0-f9df-11e7-8484-1510c7a1177e.png">

youknowriad commented:
@jasmussen I'm not able to reproduce the error. Do you know what kind of reusable block produces this error?

jasmussen commented:
I just left the office, I may not be able to reproduce at home, but it's an old installation with lots of reusable blocks created across different versions. I'm the same session I noticed a reusable block in the recent tab that was actually deleted.

mtias commented:
I was reliably reproducing this as well.

aduth commented:
Was the issue I noted in https://github.com/WordPress/gutenberg/pull/4246#pullrequestreview-86994984 addressed ?

>The placement below the inserter doesn't work particularly well when the inserter opens downward (i.e. for new posts), where I'm seeing the preview clipping the bottom of the viewport and unable to be scrolled.

youknowriad commented:
@aduth The issue was not handled but not sure what can we do about it? Any ideas? Should we show the preview on the right of the inserter instead? cc @jasmussen

jasmussen commented:
The effect of the preview when the inserter opens upwards is a blinking effect where you hover over a saved block and the preview kicks in moving the hovered block upwards, and then it goes for a loop.

Can we just show the preview _above_ the inserter, instead of below it, when the inserter opens upwards?

youknowriad commented:
@jasmussen That's another issue, we probably can do what you suggest but it doesn't solve the "scrolling" issue if there's not enough space.

jasmussen commented:
Ah, understood. 

How about we solve this in two phases. For now, we decide that this is a desktop only enhancement, with no preview on mobile — hover doesn't work too well there anyway. And given that, we show the preview on the _right_ instead of above or below. We can also make it smaller.

youknowriad committed:
Show preview on the right of the inserter

mtias commented:
I've been able to break it by creating a paragraph block, saving as reusable, selecting it on the editor canvas, then opening the inserter and hovering over it. The editor crashes with:

`TypeError: Cannot read property 'getBoundingClientRect' of null
    at Editable.getFocusPosition (http://localhost/wp-content/plugins/gutenberg/blocks/build/index.js?ver=1516926147:3023:38)`

Should be easy to reproduce.

youknowriad committed:
Format Toolbar: Make the function computing the position of the link modal generic

youknowriad commented:
Good catch @mtias this commit https://github.com/WordPress/gutenberg/pull/4246/commits/e977dc4072013e61aad6ebdaafd458f41442c32a should fix the issue. We were hardcoding selectors that do not exist in all contexts.

youknowriad commented:
This should be ready for a final review.

jasmussen commented:
I dig it! 
<img width="771" alt="screen shot 2018-01-31 at 14 00 14" src="https://user-images.githubusercontent.com/1204802/35624371-184263ae-068f-11e8-8143-5d47536a7fd5.png">

👍 👍 from me, visually

mtias commented:
@jasmussen I was seeing a few inconsistencies in the visual display of fonts, etc, that I think we should address. Might be worth applying default styles to a generic `.wp-block` selector and making sure we use that on the preview panel.

jasmussen commented:
On which previewed blocks did you see these issues?

Commit:
Polish blocks a bit.

jasmussen commented:
I see what you mean:

<img width="652" alt="screen shot 2018-02-01 at 09 17 48" src="https://user-images.githubusercontent.com/1204802/35668185-cf461522-0730-11e8-8248-5f9b3fb4e3cb.png">

I pushed a little polish:

<img width="613" alt="screen shot 2018-02-01 at 09 36 18" src="https://user-images.githubusercontent.com/1204802/35668868-882b15e0-0733-11e8-9e1c-1b26a82b3d3e.png">

<img width="639" alt="screen shot 2018-02-01 at 09 36 23" src="https://user-images.githubusercontent.com/1204802/35668871-8aaaaeb6-0733-11e8-929c-d1c114ac077a.png">

<img width="639" alt="screen shot 2018-02-01 at 09 36 28" src="https://user-images.githubusercontent.com/1204802/35668874-8dbaac8c-0733-11e8-8d78-07c72204dc2d.png">

<img width="648" alt="screen shot 2018-02-01 at 09 36 31" src="https://user-images.githubusercontent.com/1204802/35668879-8f54e5b2-0733-11e8-9224-90d03cb5be9c.png">

<img width="648" alt="screen shot 2018-02-01 at 09 36 38" src="https://user-images.githubusercontent.com/1204802/35668881-91ef751c-0733-11e8-93f6-0428c9482b46.png">

Not necessary for this PR, or perhaps at all, but this seems like an ideal use case for shadow dom.

youknowriad commented:
@jasmussen Unfortunately, we can't use shadow dom because it doesn't work on Firefox and can't be polyfilled properly.

jasmussen commented:
> Unfortunately, we can't use shadow dom because it doesn't work on Firefox and can't be polyfilled properly.

(ノಠ益ಠ)ノ彡┻━┻

No worries, it was mostly a case of "it would be nice" :)

gziolo commented:
It will look amazing as soon as we add `example` property to all blocks 🥇

youknowriad commented:
I'm thinking we can merge this right?

mtias commented:
@jasmussen we can use a class like `.wp-block` to assign the fonts, default sizes, etc, in the first place. Instead of relying on the general cascade.

youknowriad commented:
I wonder if `.wp-block` is not the generated className of the reusable block
